onnxruntime/core/providers/cpu/tensor/scatter_nd.cc

```diff
@@ -11,73 +12,87 @@ ONNX_CPU_OPERATOR_VERSIONED_KERNEL(
     11,
     12,
     KernelDefBuilder()
-        .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
-        .TypeConstraint("Tind", DataTypeImpl::GetTensorType<int64_t>()),
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
     ScatterND);
 
 ONNX_CPU_OPERATOR_KERNEL(
     ScatterND,
     13,
     KernelDefBuilder()
-        .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
-        .TypeConstraint("Tind", DataTypeImpl::GetTensorType<int64_t>()),
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
     ScatterND);
 
-template <typename Tind>
-Status ScatterNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) const {
-  const auto* input_tensor = context->Input<Tensor>(0);
-  const auto* indice_tensor = context->Input<Tensor>(1);
-  const auto* update_tensor = context->Input<Tensor>(2);
-  ORT_ENFORCE(input_tensor != nullptr);
-  ORT_ENFORCE(indice_tensor != nullptr);
-  ORT_ENFORCE(update_tensor != nullptr);
+Status ScatterNDBase::ValidateShapes(const TensorShape& input_shape,
+                                     const TensorShape& indice_shape,
+                                     const TensorShape& update_shape) const {
+  auto input_rank = input_shape.NumDimensions();
+  auto indice_rank = indice_shape.NumDimensions();
+  auto update_rank = update_shape.NumDimensions();
 
-  const auto& input_shape = input_tensor->Shape();
-  const auto& indice_shape = indice_tensor->Shape();
-  const auto& update_shape = update_tensor->Shape();
-  if (indice_shape.NumDimensions() == 0 || input_shape.NumDimensions() == 0) {
+  if (input_rank == 0 || indice_rank == 0) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                            "input tensor and indices tensor must has rank larger than 0. ",
                            "input shape: ", input_shape, ", indices shape: ", indice_shape);
   }
 
-  auto indice_rank = indice_shape.NumDimensions();
   auto last_indice_dimension = indice_shape[indice_rank - 1];
-  if (last_indice_dimension > static_cast<int64_t>(input_shape.NumDimensions())) {
+  if (last_indice_dimension > static_cast<int64_t>(input_rank)) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                            "last dimension of indices must not be larger than rank of input tensor");
   }
 
   bool is_update_shape_invalid = [&]() {
-    auto update_rank = update_shape.NumDimensions();
-    auto input_rank = input_shape.NumDimensions();
-    if (update_rank < indice_rank - 1) {
```

Comment:
Clean up: The checks here are somewhat overly complicated for the code reader (and some element of redundancy exists across all checks). Broke it up into modular pieces that will be clear to a code reader if they read the onnx spec first.